File: android/src/main/java/to/holepunch/bare/android/BareProtocol.kt
package to.holepunch.bare.android

import org.json.JSONObject
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets

object BareProtocol {
    const val VERSION: Byte = 1

    /** Mirror of `MAX_HEADER_BYTES` in `core/messages/constants.ts`. */
    const val MAX_HEADER_BYTES: Int = 0xffff

    /** Mirror of `MAX_FRAME_BYTES` in `core/messages/constants.ts`. */
    const val MAX_FRAME_BYTES: Int = 16 * 1024 * 1024

    object MessageType {
        const val ENVELOPE: Byte = 1
    }

    data class WireMessage(
        val type: Byte,
        val header: String,
        val payload: ByteArray? = null
    )

    fun buildMessage(type: Byte, headerJson: String, payload: ByteArray? = null): ByteBuffer {
        val headerBytes = headerJson.toByteArray(StandardCharsets.UTF_8)
        if (headerBytes.size > MAX_HEADER_BYTES) {
            throw IllegalArgumentException(
                "Header too large: ${headerBytes.size} bytes, maximum is $MAX_HEADER_BYTES"
            )
        }
        val payloadBytes = payload ?: ByteArray(0)

        // [version][type][headerLen(2 BE)][payloadLen(3 BE)][header][payload]:
        // the binary payload length keeps frames self-delimiting even when a
        // transport re-serializes the JSON header (mirrors core wire-codec).
        val totalLength = 7 + headerBytes.size + payloadBytes.size
        if (totalLength > MAX_FRAME_BYTES) {
            throw IllegalArgumentException(
                "Frame too large: $totalLength bytes, maximum is $MAX_FRAME_BYTES"
            )
        }
        val buffer = ByteBuffer.allocateDirect(totalLength)

        buffer.put(VERSION)
        buffer.put(type)
        buffer.putShort(headerBytes.size.toShort())
        buffer.put(((payloadBytes.size shr 16) and 0xff).toByte())
        buffer.put(((payloadBytes.size shr 8) and 0xff).toByte())
        buffer.put((payloadBytes.size and 0xff).toByte())
        buffer.put(headerBytes)
        buffer.put(payloadBytes)

        buffer.flip()
        return buffer
    }

    fun parseMessage(buffer: ByteBuffer): WireMessage? {
        if (buffer.remaining() < 7) return null
        if (buffer.remaining() > MAX_FRAME_BYTES) return null

        val version = buffer.get()
        if (version != VERSION) return null

        val type = buffer.get()
        if (type != MessageType.ENVELOPE) return null
        val headerLen = buffer.short.toInt() and 0xFFFF
        val payloadLen =
            ((buffer.get().toInt() and 0xff) shl 16) or
                ((buffer.get().toInt() and 0xff) shl 8) or
                (buffer.get().toInt() and 0xff)

        if (headerLen > MAX_HEADER_BYTES) return null
        if (buffer.remaining() < headerLen) return null
        val headerBytes = ByteArray(headerLen)
        buffer.get(headerBytes)
        val headerJson = String(headerBytes, StandardCharsets.UTF_8)

        val payload = if (buffer.remaining() >= payloadLen && payloadLen > 0) {
            val bytes = ByteArray(payloadLen)
            buffer.get(bytes)
            bytes
        } else {
            null
        }

        return WireMessage(type, headerJson, payload)
    }

    /**
     * Builds an error envelope for a frame the bridge could not deliver, keyed
     * to the original request when possible. [type] controls the wire envelope
     * type (`INVOKE_RESPONSE` for worklet-bound errors, `HOST_INVOKE_RESPONSE`
     * for host IPC errors); it defaults to `INVOKE_RESPONSE` for parity with
     * the original caller contract.
     */
    fun buildErrorResponse(headerJson: String, code: String, message: String, type: String = "INVOKE_RESPONSE"): WireMessage {
        val request = try {
            JSONObject(headerJson)
        } catch (e: Exception) {
            JSONObject()
        }
        val response = JSONObject().apply {
            put("type", type)
            put("pluginId", request.optString("pluginId"))
            put("event", request.optString("event"))
            if (request.has("requestId")) put("requestId", request.getString("requestId"))
            put(
                "error",
                JSONObject().apply {
                    put("code", code)
                    put("message", message)
                },
            )
        }
        return WireMessage(MessageType.ENVELOPE, response.toString(), null)
    }
}

File: ios/Sources/BareHost/BareProtocol.swift
import Foundation

/// Errors thrown by `BareProtocol.buildMessage` on oversized frames.
public enum BareProtocolError: Error, Equatable {
  case headerTooLarge(Int)
  case frameTooLarge(Int)
}

/**
 * Mirror of `android/.../BareProtocol.kt`: the binary wire protocol codec.
 *
 * Envelope layout (identical bytes on every transport):
 *
 * ```
 * [version:1B][type:1B][headerLen:2B BE][payloadLen:3B BE]
 * [header: UTF-8 JSON][payload: raw bytes]
 * ```
 *
 * The binary payload length keeps frames self-delimiting even when a transport
 * re-serializes the JSON header (unknown header fields may be dropped, e.g. the
 * webview→worklet relay); the raw frame bytes — including this prefix — are
 * never altered. Header parsing keeps the JSON string untouched, so known
 * fields always pass through.
 */
public enum BareProtocol {
  /// Mirror of `VERSION` in `core/messages/constants.ts`.
  public static let version: UInt8 = 1

  /// Mirror of `MAX_HEADER_BYTES` in `core/messages/constants.ts`.
  public static let maxHeaderBytes = 0xffff

  /// Mirror of `MAX_FRAME_BYTES` in `core/messages/constants.ts`.
  public static let maxFrameBytes = 16 * 1024 * 1024

  public enum MessageType {
    public static let envelope: UInt8 = 1
  }

  public struct WireMessage {
    public let type: UInt8
    public let header: String
    public let payload: Data?
  }

  /**
   * Builds a frame. Throws `BareProtocolError` when the header or total frame
   * exceeds the protocol limits — never silently truncates (see the Kotlin
   * mirror).
   */
  public static func buildMessage(
    type: UInt8,
    headerJson: String,
    payload: Data? = nil
  ) throws -> Data {
    let headerBytes = Data(headerJson.utf8)
    if headerBytes.count > maxHeaderBytes {
      throw BareProtocolError.headerTooLarge(headerBytes.count)
    }
    let payloadBytes = payload ?? Data()

    // [version][type][headerLen(2 BE)][payloadLen(3 BE)][header][payload]: the
    // binary payload length keeps frames self-delimiting even when a transport
    // re-serializes the JSON header (mirrors core wire-codec).
    let totalLength = 7 + headerBytes.count + payloadBytes.count
    if totalLength > maxFrameBytes {
      throw BareProtocolError.frameTooLarge(totalLength)
    }

    var buffer = Data()
    buffer.reserveCapacity(totalLength)
    buffer.append(version)
    buffer.append(type)
    buffer.append(UInt8((headerBytes.count >> 8) & 0xff))
    buffer.append(UInt8(headerBytes.count & 0xff))
    buffer.append(UInt8((payloadBytes.count >> 16) & 0xff))
    buffer.append(UInt8((payloadBytes.count >> 8) & 0xff))
    buffer.append(UInt8(payloadBytes.count & 0xff))
    buffer.append(headerBytes)
    buffer.append(payloadBytes)
    return buffer
  }

  /**
   * Parses one frame. Returns `nil` for any invalid frame (too short,
   * oversized, unsupported version/type, truncated or oversized header),
   * mirroring the Kotlin `parseMessage` which returns null.
   */
  public static func parseMessage(_ buffer: Data) -> WireMessage? {
    let bytes = [UInt8](buffer)
    // Minimum envelope is [version][type][headerLen:2][payloadLen:3] = 7 bytes.
    // A shorter buffer must be rejected here, before the payloadLen read below
    // indexes bytes[4..6]; otherwise a 4–6 byte frame fatally traps.
    if bytes.count < 7 { return nil }
    if bytes.count > maxFrameBytes { return nil }

    let versionByte = bytes[0]
    if versionByte != version { return nil }

    let typeByte = bytes[1]
    if typeByte != MessageType.envelope { return nil }

    let headerLen = Int(bytes[2]) << 8 | Int(bytes[3])
    if headerLen > maxHeaderBytes { return nil }
    let payloadLen =
      (Int(bytes[4]) << 16) | (Int(bytes[5]) << 8) | Int(bytes[6])
    if bytes.count < 7 + headerLen + payloadLen { return nil }

    let headerBytes = bytes[7..<(7 + headerLen)]
    guard let headerJson = String(data: Data(headerBytes), encoding: .utf8) else {
      return nil
    }

    let payload: Data? =
      payloadLen > 0 && bytes.count >= 7 + headerLen + payloadLen
      ? Data(bytes[(7 + headerLen)..<(7 + headerLen + payloadLen)])
      : nil

    return WireMessage(type: typeByte, header: headerJson, payload: payload)
  }

  /**
   * Builds an `INVOKE_RESPONSE` envelope reporting an error for a frame the
   * bridge could not deliver, keyed to the original request when possible.
   */
  public static func buildErrorResponse(
    headerJson: String,
    code: String,
    message: String
  ) -> WireMessage {
    let request =
      (try? JSONSerialization.jsonObject(with: Data(headerJson.utf8)))
      as? [String: Any] ?? [:]
    var response: [String: Any] = [
      "type": "INVOKE_RESPONSE",
      "pluginId": request["pluginId"] as? String ?? "",
      "event": request["event"] as? String ?? "",
      "error": [
        "code": code,
        "message": message,
      ],
    ]
    if let requestId = request["requestId"] as? String {
      response["requestId"] = requestId
    }
    return WireMessage(
      type: MessageType.envelope,
      header: encodeJSON(response),
      payload: nil
    )
  }

  /** Compact JSON encoding; returns `{}` if the value cannot be encoded. */
  public static func encodeJSON(_ value: Any) -> String {
    guard let data = try? JSONSerialization.data(
      withJSONObject: value,
      options: [.withoutEscapingSlashes]
    ) else {
      return "{}"
    }
    return String(data: data, encoding: .utf8) ?? "{}"
  }
}
